src/java.desktop/macosx/classes/sun/lwawt/macosx/CTextPipe.java

```diff
@@ -81,17 +81,90 @@ public void drawString(final SunGraphics2D sg2d, final String s, final double x,
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Font getSlotFont(Font font, int slot) {
+        Font2D f2d = FontUtilities.getFont2D(font);
+        if (f2d instanceof CFont) {
+            CompositeFont cf = ((CFont)f2d).getCompositeFont2D();
+            PhysicalFont pf = cf.getSlotFont(slot);
+            Font f = new Font(pf.getFontName(null),
+                              font.getStyle(), font.getSize());
+            return f;
+        }
+        return null;
+    }
+
+    private GlyphVector getGlyphVectorWithRange(final Font font, final GlyphVector gV, int start, int count) {
+        final int length = gV.getNumGlyphs();
+        int[] glyphs = new int[count];
+        for (int i = 0; i < count; i++) {
+            glyphs[i] = gV.getGlyphCode(start+i) & CompositeGlyphMapper.GLYPHMASK;
+        }
+        // Positions should be null to recalculate by native methods,
+        // if GV was segmented.
+        StandardGlyphVector sgv = new StandardGlyphVector(font,
+                                          gV.getFontRenderContext(),
+                                          glyphs,
+                                          null, // positions
+                                          null, // indices
+                                          gV.getLayoutFlags());
+        return sgv;
+    }
+
+    private int getLengthOfSameSlot(final GlyphVector gV, final int targetSlot, final int start, final int length) {
+        int count = 1;
+        for (; start + count < length; count++) {
+            int slot = (gV.getGlyphCode(start + count) &
+                        CompositeGlyphMapper.SLOTMASK) >> 24;
+            if (targetSlot != slot) {
+                break;
+            }
+        }
+        return count;
+    }
 
+    private void drawGlyphVectorImpl(final SunGraphics2D sg2d, final GlyphVector gV, final float x, final float y) {
         final long nativeStrikePtr = getNativeStrikePtr(sg2d);
         if (OSXSurfaceData.IsSimpleColor(sg2d.paint) && nativeStrikePtr != 0) {
             final OSXSurfaceData surfaceData = (OSXSurfaceData)sg2d.getSurfaceData();
             surfaceData.drawGlyphs(this, sg2d, nativeStrikePtr, gV, x, y);
         } else {
             drawGlyphVectorAsShape(sg2d, gV, x, y);
         }
+    }
+
+    public void drawGlyphVector(final SunGraphics2D sg2d, final GlyphVector gV, final float x, final float y) {
+        final Font prevFont = sg2d.getFont();
+        sg2d.setFont(gV.getFont());
+
+        if (hasSlotData(gV)) {
+            final int length = gV.getNumGlyphs();
+            float[] positions = gV.getGlyphPositions(0, length, null);
+            int start = 0;
+            while (start < length) {
+                int slot = (gV.getGlyphCode(start) &
+                            CompositeGlyphMapper.SLOTMASK) >> 24;
+                sg2d.setFont(getSlotFont(gV.getFont(), slot));
```

Comment:
Do we need to change the font for each glyph or we should do that only if the slot was changed? Or it is a common situation that each glyph will have its own slot?

Reply:
@mrserb Thank you for the comments.
Only if the slot was changed, we need to change the font. This may happen with multi-language situation.